src/Type/Php/CountCharsFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,58 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Php\PhpVersion;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+use function count;
+
+final class CountCharsFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function __construct(private PhpVersion $phpVersion)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'count_chars';
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope,
+	): ?Type
+	{
+		if (count($functionCall->getArgs()) < 1) {
+			return null;
+		}
+
+		$modeType = $scope->getType($functionCall->getArgs()[1]->value);
+
+		if (IntegerRangeType::fromInterval(0, 2)->isSuperTypeOf($modeType)->yes()) {
+			$arrayType = new ArrayType(new IntegerType(), new IntegerType());
+
+			return $this->phpVersion->throwsValueErrorForInternalFunctions()
+				? $arrayType
+				: new UnionType([$arrayType, new ConstantBooleanType(false)]);
+		}
+
+		$stringType = new StringType();
+
+		return $this->phpVersion->throwsValueErrorForInternalFunctions()
+			? $stringType
+			: new UnionType([$stringType, new ConstantBooleanType(false)]);
```

Comment:
Even in PHP 7.x and earlier, [an error would be raised before returning `false`](https://3v4l.org/ueWe1#veol), so I think wrapping the type in a [benevolent union](https://github.com/phpstan/phpstan/discussions/6440#discussioncomment-2026787) is justified.

```suggestion
			: TypeUtils::toBenevolentUnion(new UnionType([$stringType, new ConstantBooleanType(false)]));
```

Similarly for another `new UnionType`.